tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      print("FileMapping in use. Try again in a while.")
+      CloseHandle(hMapFile)
+      return
+    }
+    defer { CloseHandle(hMapFile) }
+    let FILE_MAP_ALL_ACCESS = STANDARD_RIGHTS_REQUIRED | SECTION_QUERY |
+                              SECTION_MAP_WRITE | SECTION_MAP_READ |
+                              SECTION_MAP_EXECUTE|SECTION_EXTEND_SIZE
+    let buf: LPVOID = MapViewOfFile(hMapFile,
+                                    DWORD(FILE_MAP_ALL_ACCESS),
+                                    0,
+                                    0,
+                                    SIZE_T(bufSize));
+    if buf == LPVOID(bitPattern: 0) {
+      print("MapViewOfFile failed \(GetLastError())")
+      return
+    }
+    defer { UnmapViewOfFile(buf) }
+
+    // Set up the event objects
+    let hReadEvent: HANDLE = CreateEventA(LPSECURITY_ATTRIBUTES(bitPattern: 0),
+                                          false, // Auto-reset
+                                          false, // Initial state is nonsignaled
+                                          readEventName)
+    if hReadEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hReadEvent) }
+    let hWriteEvent: HANDLE = CreateEventA(LPSECURITY_ATTRIBUTES(bitPattern: 0),
+                                           false, // Auto-reset
+                                           false, // Initial state is nonsignaled
+                                           writeEventName)
+    if hWriteEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
       return
     }
+    defer { CloseHandle(hWriteEvent) }
 
-    repeat {
-      var entry: HEAPENTRY32 = HEAPENTRY32()
-      entry.dwSize = SIZE_T(MemoryLayout<HEAPENTRY32>.size)
+    // Allocate the dll path string in the remote process.
+    guard let dllPathRemote = allocateDllPathRemote() else {
+      return
+    }
 
-      if !Heap32First(&entry, dwProcessId, heap.th32HeapID) {
-        // FIXME(compnerd) log error
-        continue
+    var hKernel32 = HMODULE(bitPattern: 0)
+    "kernel32".withCString() {
+      hKernel32 = GetModuleHandleA($0)
+    }
+
+    // Load the dll and start the heap walk
+    let hookAddr = unsafeBitCast(GetProcAddress(hKernel32, "LoadLibraryW"),
+                                 to: LPTHREAD_START_ROUTINE.self)
+    let hThread: HANDLE = CreateRemoteThread(self.process, nil, 0, hookAddr,
+                                             dllPathRemote, 0, nil)
+    if hThread == HANDLE(bitPattern: 0) {
+      print("CreateRemoteThread failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hThread) }
+
+    // The main heap iteration loop.
+    while true {
+      let wait = WaitForSingleObject(hReadEvent, waitTimeoutMs)
+      if wait != WAIT_OBJECT_0 {
+        print("WaitForSingleObject failed \(wait)")
+        return
+      }
+
+      let entryCount = bufSize/MemoryLayout<(UInt, UInt)>.size
```

Comment:
This is unsafe. You cannot marshal `(UInt, UInt)` over the wire as it is a Swift type and has no analogue in C.

Reply:
Done